stan/math/prim/mat/fun/cov_periodic.hpp

```diff
@@ -0,0 +1,144 @@
+#ifndef STAN_MATH_PRIM_MAT_FUN_COV_PERIODIC_HPP
+#define STAN_MATH_PRIM_MAT_FUN_COV_PERIODIC_HPP
+
+#include <stan/math/prim/scal/meta/return_type.hpp>
+#include <stan/math/prim/scal/err/check_size_match.hpp>
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/mat/fun/distance.hpp>
+#include <stan/math/prim/scal/err/check_not_nan.hpp>
+#include <stan/math/prim/scal/err/check_positive.hpp>
+#include <stan/math/prim/scal/fun/square.hpp>
+#include <stan/math/prim/scal/fun/inv_square.hpp>
+#include <stan/math/prim/scal/fun/inv.hpp>
+#include <vector>
+#include <cmath>
+
+namespace stan {
+  namespace math {
+
+    /**
+     * Returns a periodic covariance matrix \f$ \mathbf{K} \f$ using the input \f$ \mathbf{X} \f$.
+     * The elements of \f$ \mathbf{K} \f$ are defined as
+     * \f$ \mathbf{K}_{ij} = k(\mathbf{X}_i,\mathbf{X}_j), \f$ where
+     * \f$ \mathbf{X}_i \f$ is the \f$i\f$-th row of \f$ \mathbf{X} \f$ and \n
+     * \f$ k(\mathbf{x},\mathbf{x}^\prime) =
+     * \sigma^2 \exp\left(-\frac{2\sin^2(\pi |\mathbf{x}-\mathbf{x}^\prime|/p)}{\ell^2}\right), \f$
+     * \n
+     * where \f$ \sigma^2 \f$, \f$ \ell \f$ and \f$ p \f$ are the signal variance, length-scale and period.
+     *
+     * @tparam T_x type of std::vector elements of x.
+     *   T_x can be a scalar, an Eigen::Vector, or an Eigen::RowVector.
+     * @tparam T_sigma type of sigma
+     * @tparam T_l type of length-scale
+     * @tparam T_p type of period
+     *
+     * @param x std::vector of input elements.
+     *   This function assumes that all elements of x have the same size.
+     * @param sigma standard deviation of the signal
+     * @param l length-scale
+     * @param p period
+     * @return periodic covariance matrix
+     * @throw std::domain_error if sigma <= 0, l <= 0, p <= 0 or
+     *   x is nan or infinite
+     */
+    template<typename T_x, typename T_sigma, typename T_l, typename T_p>
+    inline typename
+    Eigen::Matrix<typename stan::return_type<T_x, T_sigma, T_l, T_p>::type,
+                  Eigen::Dynamic, Eigen::Dynamic>
+    cov_periodic(const std::vector<T_x>& x, const T_sigma& sigma, const T_l& l,
+                 const T_p& p) {
+      using std::exp;
+      check_positive("cov_periodic", "signal standard deviation", sigma);
+      check_positive("cov_periodic", "length-scale", l);
+      check_positive("cov_periodic", "period", p);
+      for (size_t n = 0; n < x.size(); ++n)
+        check_not_nan("cov_periodic", "x", x[n]);
+
+      Eigen::Matrix<typename stan::return_type<T_x, T_sigma, T_l, T_p>::type,
+                    Eigen::Dynamic, Eigen::Dynamic> cov(x.size(), x.size());
+
+      size_t x_size = x.size();
+      if (x_size == 0)
+        return cov;
+
+      T_sigma sigma_sq = square(sigma);
+      T_l neg_two_inv_l_sq = -2.0 * inv_square(l);
+      T_p pi_div_p = M_PI * inv(p);
```

Comment:
This'd be better as `M_PI / p` because it's fewer autodiff ops.  `MPI * inv(p)` needs a node for `inv(p)` and then one for the multiplication, whereas the direct division is more compact and hence faster.

Reply:
done